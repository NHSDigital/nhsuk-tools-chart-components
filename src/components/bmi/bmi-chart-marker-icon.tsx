import React, { JSX } from 'react';

export function BmiChartMarkerIcon(): JSX.Element {
  return (
    <svg viewBox="1 1 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
      <g clipPath="url(#clip0_1346_102759)">
        <path
          xmlns="http://www.w3.org/2000/svg"
          d="M1.93209 4.05025L1.57854 4.40381L1.93209 4.75736L6.17473 9L1.93209 13.2426L1.57854 13.5962L1.93209 13.9497L4.05341 16.0711L4.40697 16.4246L4.76052 16.0711L9.00316 11.8284L13.2458 16.0711L13.5994 16.4246L13.9529 16.0711L16.0742 13.9497L16.4278 13.5962L16.0742 13.2426L11.8316 9L16.0742 4.75736L16.4278 4.40381L16.0742 4.05025L13.9529 1.92893L13.5994 1.57538L13.2458 1.92893L9.00316 6.17157L4.76052 1.92893L4.40697 1.57538L4.05341 1.92893L1.93209 4.05025Z"
          fill="black"
          stroke="white"
        />
      </g>
    </svg>
  );
}
