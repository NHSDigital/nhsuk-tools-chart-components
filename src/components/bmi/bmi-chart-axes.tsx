import React, { JSX } from 'react';
import { BmiResultChart } from '@models/bmi-chart.model';
import { BmiClassificationLabel } from '@custom-types/bmi.types';
import './styles/bmi-chart-axes.module.scss';

type BmitGraphAxisMarkerProps = {
  markerText: string;
  offsetPercent: number;
};

type BmitGraphAxesProps = {
  model: BmiResultChart;
};

function AxisMarker({
  markerText,
  offsetPercent,
}: Readonly<BmitGraphAxisMarkerProps>): JSX.Element {
  return (
    <div
      className="nhsuk-bmi-axis-marker"
      style={{ left: `${offsetPercent}%` }}>
      <div className={`${'nhsuk-bmi-axis-marker-label'}`}>{markerText}</div>
    </div>
  );
}

export function BmiChartAxes({
  model,
}: Readonly<BmitGraphAxesProps>): JSX.Element {
  return (
    <>
      <AxisMarker
        markerText={model.bounds.upperBounds[
          BmiClassificationLabel.Underweight
        ].toString()}
        offsetPercent={model.classificationPercentage(
          BmiClassificationLabel.Underweight
        )}
      />
      <AxisMarker
        markerText={model.bounds.upperBounds[
          BmiClassificationLabel.Healthy
        ].toString()}
        offsetPercent={model.classificationPercentage(
          BmiClassificationLabel.Healthy
        )}
      />
      <AxisMarker
        markerText={model.bounds.upperBounds[
          BmiClassificationLabel.Overweight
        ].toString()}
        offsetPercent={model.classificationPercentage(
          BmiClassificationLabel.Overweight
        )}
      />
    </>
  );
}
