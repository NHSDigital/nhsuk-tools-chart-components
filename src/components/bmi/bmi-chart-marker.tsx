import React, { JSX } from 'react';
import { BmiResultChart } from '@models/bmi-chart.model';
import { BmiChartMarkerIcon } from './bmi-chart-marker-icon';
import './styles/bmi-chart-marker.module.scss';

export function BmiChartMarker({
  bmi,
  model,
}: Readonly<{
  bmi: number;
  model: BmiResultChart;
}>): JSX.Element {
  const offset = model.offsetPercentage(bmi);

  let roundedOffset = 0;
  if (offset > 0) {
    const roundingPrecision = 1;
    const multiplier = 10 ** roundingPrecision;

    roundedOffset = Math.round(offset * multiplier) / multiplier;
  }

  return (
    <div
      className="nhsuk-bmi-chart-marker"
      style={{ left: `${roundedOffset}%` }}>
      <BmiChartMarkerIcon />
    </div>
  );
}
