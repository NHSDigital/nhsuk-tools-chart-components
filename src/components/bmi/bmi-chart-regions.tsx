import React, { JSX } from 'react';
import { BmiResultChart } from '@models/bmi-chart.model';
import {
  BmiClassificationLabel,
  BmiClassificationLevelLabel,
} from '@custom-types/bmi.types';
import './styles/bmi-chart-regions.module.scss';

function RegionForClassification({
  classification,
  model,
}: Readonly<{
  classification: BmiClassificationLevelLabel;
  model: BmiResultChart;
}>): JSX.Element {
  const width = model.classificationPercentage(classification);

  return (
    <div
      className={`nhsuk-bmi-classification-${classification}`}
      style={{ width: `${width}%` }}
    />
  );
}

export function BmiChartRegions({
  model,
}: Readonly<{
  model: BmiResultChart;
}>): JSX.Element {
  return (
    <>
      <RegionForClassification
        classification={BmiClassificationLabel.Underweight}
        model={model}
      />
      <RegionForClassification
        classification={BmiClassificationLabel.Healthy}
        model={model}
      />
      <RegionForClassification
        classification={BmiClassificationLabel.Overweight}
        model={model}
      />
      <RegionForClassification
        classification={BmiClassificationLabel.Obese}
        model={model}
      />
    </>
  );
}
