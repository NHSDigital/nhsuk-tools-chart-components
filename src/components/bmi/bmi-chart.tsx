import React, { JSX } from 'react';
import { BmiResultChart } from '@models/bmi-chart.model';
import { BmiClassificationBounds } from '@custom-types/bmi.types';
import { BmiChartMarker } from './bmi-chart-marker';
import { BmiChartRegions } from './bmi-chart-regions';
import { BmiChartAxes } from './bmi-chart-axes';
import './styles/bmi-chart.module.scss';

/**
 * The properties for @BmiChart component
 *
 * @param {classificationBounds} classificationBounds -  The body mass index classification ranges
 * @param {number} bmi - The body mass index value
 */
export type BmiChartProps = {
  classificationBounds: BmiClassificationBounds;
  bmi: number;
};

/**
 * A component to display the body mass index chart
 *
 * @typedef {object} BmiChartProps
 * @property {classificationBounds} classificationBounds -  The body mass index classification ranges
 * @property {number} bmi - The body mass index value
 *
 * @param {BmiChartProps} props
 * @returns {JSX.Element}
 */
export function BmiChart({
  classificationBounds,
  bmi,
}: Readonly<BmiChartProps>): JSX.Element {
  const chartModel = new BmiResultChart(classificationBounds);

  return (
    <div className="nhsuk-bmi-graph" aria-hidden="true">
      <BmiChartAxes model={chartModel} />
      <BmiChartMarker bmi={bmi} model={chartModel} />
      <BmiChartRegions model={chartModel} />
    </div>
  );
}
