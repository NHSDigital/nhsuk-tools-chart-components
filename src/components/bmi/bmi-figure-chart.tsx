import React, { JSX } from 'react';
import {
  BmiClassificationBounds,
  BmiLegendKeys,
} from '@custom-types/bmi.types';
import { ChartLegend, ChartLegendKey } from '@components/chart-legend';
import { BmiChart } from './bmi-chart';
import { BmiChartMarkerIcon } from './bmi-chart-marker-icon';
import './styles/bmi-chart-card.module.scss';

const populateBmiChartLegendItems = (
  legendKeys: BmiLegendKeys,
  legendMarkerText: string
): ChartLegendKey[] =>
  Object.entries(legendKeys)
    .map(
      ([classification, label]): ChartLegendKey => ({
        text: label,
        className: `nhsuk-bmi-legend-key-classification-${classification}`,
      })
    )
    .concat({
      text: legendMarkerText,
      className: 'nhsuk-bmi-legend-key-marker',
      element: <BmiChartMarkerIcon />,
    });

/**
 * The properties for @BmiChartFigure component
 *
 * @param {classificationBounds} classificationBounds -  The body mass index classification ranges
 * @param {number} bmi - The body mass index value
 * @param {string} ariaLabel - The aria label for the chart figure
 * @param {BmiLegendKeys} legendKeys - The chart legend items
 * @param {string} legendMarkerText The chart legend marker text
 */
export type BmiChartFigureProps = {
  classificationBounds: BmiClassificationBounds;
  bmi: number;
  ariaLabel: string;
  legendKeys: BmiLegendKeys;
  legendMarkerText: string;
};

/**
 * A component to display the body mass index chart figure
 *
 * @typedef {object} BmiChartFigureProps
 * @property {classificationBounds} classificationBounds -  The body mass index classification ranges
 * @property {number} bmi - The body mass index value
 * @property {string} ariaLabel - The aria label for the chart figure
 * @property {BmiLegendKeys} legendKeys - The chart legend items
 * @property {string} legendMarkerText The chart legend marker text
 *
 * @param {BmiChartFigureProps} props
 * @returns {JSX.Element}
 */
export function BmiChartFigure({
  classificationBounds,
  bmi,
  ariaLabel,
  legendKeys,
  legendMarkerText,
}: Readonly<BmiChartFigureProps>): JSX.Element {
  const bmiLegendKeys = populateBmiChartLegendItems(
    legendKeys,
    legendMarkerText
  );

  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <BmiChart classificationBounds={classificationBounds} bmi={bmi} />

      <ChartLegend legendKeys={bmiLegendKeys} borderWrapper />
    </figure>
  );
}
