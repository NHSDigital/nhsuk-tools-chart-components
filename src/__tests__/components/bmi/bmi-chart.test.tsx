import { BmiClassificationBounds } from '@custom-types/bmi.types';
import { render } from '@testing-library/react';
import { mockedBmiClassificationBounds } from '@tests/__helpers__/mocks/bmi-results';
import React, { JSX } from 'react';
import { BmiChart } from '@components/bmi/bmi-chart';
import '@testing-library/jest-dom';

function ResultBmiResultChartTestApp({
  classificationBounds = mockedBmiClassificationBounds,
  bmi = 0,
}: Readonly<{
  classificationBounds?: BmiClassificationBounds;
  bmi?: number;
}>): JSX.Element {
  return <BmiChart classificationBounds={classificationBounds} bmi={bmi} />;
}

describe('The BMI result chart component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<ResultBmiResultChartTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders the chart axis ', () => {
    const { container } = render(<ResultBmiResultChartTestApp />);

    const axisElements = container?.querySelectorAll('.nhsuk-bmi-axis-marker');

    expect(axisElements.length).toBe(3);
    const underweightAxisEl = axisElements[0] as HTMLElement;
    const healthyAxisEl = axisElements[1] as HTMLElement;
    const overweightAxisEl = axisElements[2] as HTMLElement;

    expect(underweightAxisEl.style.left).toBe('4.444444444444445%');
    expect(healthyAxisEl.style.left).toBe('33.33333333333333%');
    expect(overweightAxisEl.style.left).toBe('55.55555555555556%');

    const axisLabelElements = container?.querySelectorAll(
      '.nhsuk-bmi-axis-marker-label'
    );

    expect(axisLabelElements.length).toBe(3);
    const underweightAxisLabelEl = axisLabelElements[0];
    const healthyAxisLabelEl = axisLabelElements[1];
    const overweightAxisLabelEl = axisLabelElements[2];

    expect(underweightAxisLabelEl.innerHTML).toBe('18.5');
    expect(healthyAxisLabelEl.innerHTML).toBe('25');
    expect(overweightAxisLabelEl.innerHTML).toBe('30');
  });

  it('Renders the chart regions', () => {
    const { container } = render(<ResultBmiResultChartTestApp />);

    const underweightRegionEl = container?.querySelector(
      '.nhsuk-bmi-classification-underweight'
    ) as HTMLElement;
    expect(underweightRegionEl).not.toBeNull();
    expect(underweightRegionEl.style.width).toBe('4.444444444444445%');

    const healthyRegionEl = container?.querySelector(
      '.nhsuk-bmi-classification-healthy'
    ) as HTMLElement;
    expect(healthyRegionEl).not.toBeNull();
    expect(healthyRegionEl.style.width).toBe('33.33333333333333%');

    const overweightRegionEl = container?.querySelector(
      '.nhsuk-bmi-classification-overweight'
    ) as HTMLElement;
    expect(overweightRegionEl).not.toBeNull();
    expect(overweightRegionEl.style.width).toBe('55.55555555555556%');

    const obeseRegionEl = container?.querySelector(
      '.nhsuk-bmi-classification-obese'
    ) as HTMLElement;
    expect(obeseRegionEl).not.toBeNull();
    expect(obeseRegionEl.style.width).toBe('100%');
  });

  it.each`
    bmiValue | expectedPosition
    ${0}     | ${0}
    ${12}    | ${0}
    ${30}    | ${55.6}
    ${45}    | ${122.2}
    ${67}    | ${220}
    ${89}    | ${317.8}
    ${92}    | ${331.1}
    ${98}    | ${357.8}
    ${100}   | ${366.7}
  `(
    'expect the chart to contain a chart marker positioned at $expectedPosition',
    async ({
      bmiValue,
      expectedPosition,
    }: {
      bmiValue: number;
      expectedPosition: number;
    }) => {
      const { container } = render(
        <ResultBmiResultChartTestApp bmi={bmiValue} />
      );

      const markerEl = container?.querySelector(
        '.nhsuk-bmi-chart-marker'
      ) as HTMLElement;
      expect(markerEl).not.toBeNull();
      expect(markerEl.style.left).toBe(`${expectedPosition}%`);
    }
  );
});
