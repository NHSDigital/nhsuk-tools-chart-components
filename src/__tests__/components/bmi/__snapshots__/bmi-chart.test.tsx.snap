// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The BMI result chart component Renders with the expected content 1`] = `
<div>
  <div
    aria-hidden="true"
    class="nhsuk-bmi-graph"
  >
    <div
      class="nhsuk-bmi-axis-marker"
      style="left: 4.444444444444445%;"
    >
      <div
        class="nhsuk-bmi-axis-marker-label"
      >
        18.5
      </div>
    </div>
    <div
      class="nhsuk-bmi-axis-marker"
      style="left: 33.33333333333333%;"
    >
      <div
        class="nhsuk-bmi-axis-marker-label"
      >
        25
      </div>
    </div>
    <div
      class="nhsuk-bmi-axis-marker"
      style="left: 55.55555555555556%;"
    >
      <div
        class="nhsuk-bmi-axis-marker-label"
      >
        30
      </div>
    </div>
    <div
      class="nhsuk-bmi-chart-marker"
      style="left: 0%;"
    >
      <svg
        fill="none"
        viewBox="1 1 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          clip-path="url(#clip0_1346_102759)"
        >
          <path
            d="M1.93209 4.05025L1.57854 4.40381L1.93209 4.75736L6.17473 9L1.93209 13.2426L1.57854 13.5962L1.93209 13.9497L4.05341 16.0711L4.40697 16.4246L4.76052 16.0711L9.00316 11.8284L13.2458 16.0711L13.5994 16.4246L13.9529 16.0711L16.0742 13.9497L16.4278 13.5962L16.0742 13.2426L11.8316 9L16.0742 4.75736L16.4278 4.40381L16.0742 4.05025L13.9529 1.92893L13.5994 1.57538L13.2458 1.92893L9.00316 6.17157L4.76052 1.92893L4.40697 1.57538L4.05341 1.92893L1.93209 4.05025Z"
            fill="black"
            stroke="white"
            xmlns="http://www.w3.org/2000/svg"
          />
        </g>
      </svg>
    </div>
    <div
      class="nhsuk-bmi-classification-underweight"
      style="width: 4.444444444444445%;"
    />
    <div
      class="nhsuk-bmi-classification-healthy"
      style="width: 33.33333333333333%;"
    />
    <div
      class="nhsuk-bmi-classification-overweight"
      style="width: 55.55555555555556%;"
    />
    <div
      class="nhsuk-bmi-classification-obese"
      style="width: 100%;"
    />
  </div>
</div>
`;
